rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeper.java

```diff
@@ -224,42 +227,61 @@ public Map getCacheStats()
     //using localTransferServer.
     // If the data is not already cached, remoteReadRequest for that block is sent and data is cached.
 
-    //TODO : buffer is initialized everytime readData is called and it contains garbage value which is not required. Also, getCacheStatus is called twice.
+    //TODO : buffer is initialized everytime readData is called and it contains garbage value which is not required.
+     // We cannot make it static because the variable is used in remoteReadRequestChain to write (cache) data to files.
+    // So, it has to store the correct value in each thread. getCacheStatus is called twice.
     @Override
-    public boolean readData(String path, long offset, int length, long fileSize, long lastModified, int clusterType)
+    public boolean readData(String remotePath, long offset, int length, long fileSize, long lastModified, int clusterType)
             throws TException
     {
         int blockSize = CacheConfig.getBlockSize(conf);
         byte[] buffer = new byte[blockSize];
         BookKeeperFactory bookKeeperFactory = new BookKeeperFactory(this);
-        CachingNativeS3FileSystem fs = null;
+        FileSystem fs = null;
         FSDataInputStream inputStream = null;
+        Path path = new Path(remotePath);
         long startBlock = offset / blockSize;
         long endBlock = ((offset + (length - 1)) / CacheConfig.getBlockSize(conf)) + 1;
         try {
             int idx = 0;
-            List<BlockLocation> blockLocations = getCacheStatus(path, fileSize, lastModified, startBlock, endBlock, clusterType);
+            List<BlockLocation> blockLocations = getCacheStatus(remotePath, fileSize, lastModified, startBlock, endBlock, clusterType);
 
             for (int blockNum = (int) startBlock; blockNum < endBlock; blockNum++, idx++) {
                 int readStart = blockNum * blockSize;
+                log.debug(" blockLocation is: " + blockLocations.get(idx).getLocation());
                 if (blockLocations.get(idx).getLocation() != Location.CACHED) {
                     if (fs == null) {
-                        fs = new CachingNativeS3FileSystem(bookKeeperFactory, new Path(path), conf);
-                        inputStream = fs.open(new Path(path), blockSize);
```

Comment:
This needs a change in README. Need to inform ppl that the hadoop config while starting BookKeeper should be set correctly to return CachingFileSystem class for remote paths.